src/pallene/flow.lua

```diff
@@ -0,0 +1,309 @@
+    Build = {
+        "start",   -- set of values when we start analysing the block
+        "finish",  -- set of values when we finish analysing the block,
+        "kill",    -- set of values
+        "gen",     -- set of values
+    },
+
+    Apply = {
+        "set",     -- set of values
+    },
+})
+
+function flow.FlowInfo(order, process_cmd, init_start)
+    return {
+        order = order,              -- flow.Order
+        process_cmd = process_cmd,  -- function
+        init_start = init_start,    -- function
+    }
+end
+
+local function copy_set(S)
+    local new_set = {}
+    for v,_ in pairs(S) do
+        new_set[v] = true
+    end
+    return new_set
+end
+
+local function apply_gen_kill_sets(flow_state)
+    local start = flow_state.start
+    local finish = flow_state.finish
+    local gen = flow_state.gen
+    local kill = flow_state.kill
+    local in_changed = false
+
+    for v, _ in pairs(start) do
+        local val = true
+        if kill[v] then
+            val = nil
+        end
+        local previous_val = finish[v]
+        local new_val = previous_val or val
+        finish[v] = new_val
+        in_changed = in_changed or (previous_val ~= new_val)
+    end
+
+    for v, g in pairs(gen) do
+        assert(not (g and kill[v]), "gen and kill can't both be true")
+        local previous_val = finish[v]
+        local new_val = true
+        finish[v] = new_val
+        in_changed = in_changed or (previous_val ~= new_val)
+    end
+
+    for v, _ in pairs(finish) do
+        if not start[v] and not gen[v] then
+            finish[v] = nil
+            in_changed = true
+        end
+    end
+
+    return in_changed
+end
+
+local function clear_set(S)
+    for v,_ in pairs(S) do
+        S[v] = nil
+    end
+end
+
+local function merge_sets(state_list, src_indices, dest_index)
+    local dest = state_list[dest_index].start
+    clear_set(dest)
+    for _,src_i in ipairs(src_indices) do
+        local src = state_list[src_i].finish
+        for v, _ in pairs(src) do
+            dest[v] = true
+        end
+    end
+end
+
+local function make_state_list(block_list, flow_info)
+    local state_list = {}
+    local order = flow_info.order._tag
+    for block_i, block in ipairs(block_list) do
+        local state = flow.FlowState.Build({},{},{},{})
+        state_list[block_i] = state
+        flow_info.init_start(state.start, block_i)
+        if order == "flow.Order.Forward" then
+            for cmd_i = 1, #block.cmds do
+                flow_info.process_cmd(state, block_i, cmd_i)
+            end
+        elseif order == "flow.Order.Backwards" then
+            for cmd_i = #block.cmds, 1, -1  do
+                flow_info.process_cmd(state, block_i, cmd_i)
+            end
+        else
+            tagged_union.error(order)
+        end
+    end
+    return state_list
+end
+
+function flow.flow_analysis(block_list, flow_info)
+
+    local state_list = make_state_list(block_list, flow_info)
```

Comment:
Is there one state per block? If so, we could make that more obvious by calling it "state_of_block" instead of "state_list" (which in theory could have any number of states, in any order)